### PR TITLE
Thin pool support

### PR DESCRIPTION
**Pull request by japokorn**
Storage role support for thin provisioning.
Volumes under LVM pool now can use three new options:
 - 'thin' - type bool
          - setting to True puts the volume into a thin pool
 - 'thin_pool_name' - type str
          - specifies the name of the thin pool for this volume
          - selects thin pool with the same name if it exists
          - tries to create new thin pool if it does not
          - if omitted and no existing thin pool is present, the name is generated automatically
          - if omitted and exactly one thin pool is present, the volume is put into existing thin pool
          - if omitted and more than one thin pool is present, exception is raised
 - 'thin_pool_size' - type Size
          - specifeis size of the thin pool

### CODE REVIEW

File: library/blivet.py [hunk 1]
```diff
@@ -791,29 +791,53 @@ def _get_params_lvmcache(self):
 
         return dict(cache_request=cache_request)
 
+    def _get_parent_thinpool_by_name(self, thin_pool_name):
+
+        if thin_pool_name is None:
+            if len(self._blivet_pool._thinpools) != 1:
+                raise BlivetAnsibleError("cannot determine thinpool for assigning volume '%s' " % self._volume.get('name'))
+            else:
+                return self._blivet_pool._thinpools[0]
```
Comment (richm) [inline library/blivet.py:800]: in the case where `len(self._blivet_pool._thinpools) == 1`, is it always true that `thin_pool_name == self._blivet_pool._thinpools[0].name`?

**Comment by richm** [comment]
There's is some quite complex python code added . . .
For the blivet related code - @vojtechtrefny could you take a look?
For the validation code - would be really nice to have some sort of unit tests for this (and yes, I realize we have very little unit testing code as it is . . .)

File: library/blivet.py (hunk 1 above)
Comment (japokorn) [reply library/blivet.py:800]: Yes. BlivetLVMVolume._get_parent_thinpool_by_name is called from BlivetLVMVolume._create under condition that thin pool creation is requested.
But before BlivetLVMVolume._create is called BlivetLVMPool._manage_thin_pools is executed which creates thin pools (and/or finds the existing ones).

**Comment by japokorn** [comment]
@richm Makes sense. At this moment argument validation testing relies on the fact that our test suite uses wide variety of parameter combinations, hence it is at least 'somewhat' tested.
So my plan is to focus on extending 'regular' thin pool tests first, then move on to argument validation unit tests.

File: library/blivet.py [hunk 2]
```diff
@@ -1265,18 +1289,98 @@ def _manage_encryption(self, members):
 
         return managed_members
 
+    def _manage_thin_pools(self, pool_device):
+
+        existing_thinlvs_obj = [d for d in pool_device.children if d.type == 'lvmthinpool']
```
Comment (vojtechtrefny) [inline library/blivet.py:1294]: You can use `is_thin_pool` instead of `d.type == 'lvmthinpool'`.

File: module_utils/storage_lsr/argument_validator.py [hunk 3]
```diff
@@ -7,6 +7,26 @@
 # pylint: disable=undefined-variable
 STRING_TYPE = str if sys.version_info.major == 3 else basestring  # noqa:F821
 
+BOOLEANS_TRUE = ['y', 'yes', 'on', '1', 'true', 't', 1, 1.0, True]
```
Comment (vojtechtrefny) [inline module_utils/storage_lsr/argument_validator.py:10]: I think this should be a separate commit with some comment, this doesn't look related to the thin provisioning support.

Comment (japokorn) [reply module_utils/storage_lsr/argument_validator.py:10]: Fixed.

File: library/blivet.py [hunk 4]
```diff
@@ -1265,18 +1289,98 @@ def _manage_encryption(self, members):
...
+                        reserved_space += Size(thin_size)
+            else:
+                # regular LV just take its size
+                vol_size = volume.get('size')
+                if vol_size is None:
+                    auto_size_dev_count += 1
+                else:
+                    reserved_space += Size(vol_size)
+
+        # Thin metadata will take 20% of VG space. At least 1GiB, at most 100GiB
```
[the remark this answers is not in the document]
Comment (japokorn) [reply library/blivet.py]: Thanks for clarification. Fixed.

File: library/blivet.py [hunk 5]
```diff
@@ -791,29 +791,53 @@ def _get_params_lvmcache(self):
...
+
+        if thin_pool_name is None:
+            if len(self._blivet_pool._thinpools) != 1:
+                raise BlivetAnsibleError("cannot determine thinpool for assigning volume '%s' " % self._volume.get('name'))
+            else:
+                return self._blivet_pool._thinpools[0]
+
+        for thinp in self._blivet_pool._thinpools:
+            # lvm concatenates the name as <vg name>-<thin pool name>
+            if self._blivet_pool._device.name + '-' + thin_pool_name == thinp.name:
```
[the remark this answers is not in the document]
Comment (japokorn) [reply library/blivet.py]: Thanks. Fixed.

File: library/blivet.py [hunk 6]
```diff
@@ -153,6 +152,7 @@
 
 
 MAX_TRIM_PERCENT = 2
+DEFAULT_THPOOL_RESERVE = dict(percent=20, min=Size("1 GiB"), max=Size("100 GiB"))
```
[the remark this answers is not in the document]
Comment (japokorn) [reply library/blivet.py]: It was a remnant from debugging. Thanks for noticing. Fixed.

File: library/blivet.py [hunk 7]
```diff
@@ -1538,6 +1642,9 @@ def run_module():
                                                           raid_level=dict(type='str'),
                                                           size=dict(type='str'),
                                                           state=dict(type='str', default='present', choices=['present', 'absent']),
+                                                          thin_pool_name=dict(type='str'),
+                                                          thin_pool_size=dict(type='str'),
+                                                          thin=dict(type='bool', default=False),
```
Comment (richm) [inline library/blivet.py:1647]: You'll have to rebase this on top of my recent change

**Comment by richm** [comment]
centos-7 tests are failing due to missing libraries:
```
** (process:2499): [1;33mWARNING[0m **: [34m13:18:30.150[0m: failed to load module btrfs: libbd_btrfs.so.2: cannot open shared object file: No such file or directory

** (process:2499): [1;33mWARNING[0m **: [34m13:18:30.150[0m: failed to load module loop: libbd_loop.so.2: cannot open shared object file: No such file or directory

** (process:2499): [1;33mWARNING[0m **: [34m13:18:30.151[0m: failed to load module mpath: libbd_mpath.so.2: cannot open shared object file: No such file or directory

** (process:2499): [1;33mWARNING[0m **: [34m13:18:30.152[0m: failed to load module nvdimm: libbd_nvdimm.so.2: cannot open shared object file: No such file or directory
```

File: library/blivet.py [hunk 8]
```diff
@@ -1265,18 +1289,98 @@ def _manage_encryption(self, members):
...
+                        auto_size_dev_count += 1
+                    else:
+                        reserved_space += Size(thin_size)
+            else:
+                # regular LV just take its size
+                vol_size = volume.get('size')
+                if vol_size is None:
+                    auto_size_dev_count += 1
+                else:
+                    reserved_space += Size(vol_size)
```
Comment (richm) [inline library/blivet.py:1331]: This is causing the wrong error message to be thrown.  The tests are expecting an error from `_get_size()` like `"invalid percentage '2x%' size specified for volume '%s'"`.  Instead, the error message is more cryptic:
```
ValueError: Failed to recognize unit from the spec: 2x%
```
I think you are not supposed to use `Size()` directly unless you are sure the value is not a percentage.  I think you are supposed to use `Size()` directly unless you are sure the value is not a percentage.  I think you are supposed to use `self._get_size()`.  Unfortunately, I'm not sure which object to use for the `_get_size` method - the code is in `class BlivetLVMPool(BlivetPool)` but only `BlivetVolume` has the `_get_size` method.

Alternately you could change the test to look for the above error message, but IMO that isn't a good error message.

Comment (japokorn) [reply library/blivet.py]: Thank you. Fixed.

File: library/blivet.py [hunk 9]
```diff
@@ -1265,18 +1289,114 @@ def _manage_encryption(self, members):
 
         return managed_members
 
+    def _manage_thin_pools(self, pool_device):
+
+        def get_vol_size(volume):
```
Comment (richm) [inline library/blivet.py:1294]: The parameter name should be `vol_size`, or `vol_size_str` if you think that would be more descriptive.

File: library/blivet.py [hunk 10]
```diff
@@ -1265,18 +1289,114 @@ def _manage_encryption(self, members):
...
+            if isinstance(volume, str) and '%' in volume:
+                try:
+                    percentage = int(volume[:-1].strip())
+                except ValueError:
+                    raise BlivetAnsibleError("invalid percentage '%s' size specified for volume in pool '%s'" % (pool_device.name))
```
Comment (richm) [inline library/blivet.py:1299]: Error here - `TypeError: not enough arguments for format string` - you are missing the volume size str - so
```
                     raise BlivetAnsibleError("invalid percentage '%s' size specified for volume in pool '%s'" % (vol_size, pool_device.name))
```

**Comment by richm** [comment]
any updates planned?

**Comment by richm** [comment]
this lgtm - @vojtechtrefny ?

The test errors are due to CI issues
